lib/dao.js

```diff
@@ -2388,15 +2390,15 @@ DataAccessObject.prototype.updateAttributes = function updateAttributes(data, op
             });
           }
 
-          context = {
+          var ctx = {
             Model: Model,
             where: byIdQuery(Model, getIdValue(Model, inst)).where,
-            data: data,
+            data: context.data,
             currentInstance: inst,
             hookState: hookState,
             options: options
           };
-          Model.notifyObserversOf('persist', context, function(err) {
+          Model.notifyObserversOf('persist', ctx, function(err) {
             if (connector.updateAttributes.length === 5) {
               connector.updateAttributes(model, getIdValue(inst.constructor, inst),
                 inst.constructor._forDB(typedData), options, updateAttributesCallback);
```

Comment:
> verify that user.extra is set, user.address is not set

Check the data in the database too. If I am reading the code correctly, it will apply hook updates to the instance returned to callback, but it won't apply them in the database.